server/src/internalClusterTest/java/org/opensearch/action/bulk/AppendOnlyIndicesIT.java

```diff
@@ -0,0 +1,170 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.action.bulk;
+
+import org.opensearch.action.admin.cluster.node.stats.NodeStats;
+import org.opensearch.action.admin.cluster.node.stats.NodesStatsResponse;
+import org.opensearch.client.Client;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.core.xcontent.XContentBuilder;
+import org.opensearch.ingest.IngestTestPlugin;
+import org.opensearch.plugins.Plugin;
+import org.opensearch.test.OpenSearchIntegTestCase;
+import org.opensearch.test.transport.MockTransportService;
+import org.opensearch.transport.ConnectTransportException;
+import org.opensearch.transport.TransportService;
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.stream.Collectors;
+
+import static org.opensearch.common.xcontent.XContentFactory.jsonBuilder;
+import static org.opensearch.test.hamcrest.OpenSearchAssertions.assertAcked;
+import static org.hamcrest.CoreMatchers.equalTo;
+import static org.hamcrest.Matchers.containsString;
+
+public class AppendOnlyIndicesIT extends OpenSearchIntegTestCase {
+
+    @Override
+    protected Collection<Class<? extends Plugin>> nodePlugins() {
+        return Arrays.asList(IngestTestPlugin.class, MockTransportService.TestPlugin.class);
+    }
+
+    public void testIndexDocumentWithACustomDocIdForAppendOnlyIndices() throws Exception {
+        Client client = internalCluster().coordOnlyNodeClient();
+        assertAcked(
+            client().admin()
+                .indices()
+                .prepareCreate("index")
+                .setSettings(
+                    Settings.builder()
+                        .put("index.number_of_replicas", 0)
+                        .put("index.number_of_shards", 1)
+                        .put("index.append_only.enabled", true)
+                )
+        );
+        ensureGreen("index");
+
+        BulkRequestBuilder bulkBuilder = client.prepareBulk();
+
+        XContentBuilder doc = null;
+        doc = jsonBuilder().startObject().field("foo", "bar").endObject();
+        bulkBuilder.add(client.prepareIndex("index").setId(Integer.toString(0)).setSource(doc));
+
+        BulkResponse response = bulkBuilder.get();
+        assertThat(response.getItems()[0].getFailureMessage(), containsString("Operation [INDEX] is not allowed with a custom doc id 0"));
```

Comment:
when is the first bulk for doc id 0

Reply:
This is to validate scenario that indexing operation is blocked when custom document is passed for append only indices. So indexing requests fails at the first request itself.